src/config.rs

```diff
@@ -413,93 +413,104 @@ fn auth_from_data(
+                                "Invalid certificate data, could not base64 decode: {}",
                                 e.description()
                             );
                         }
+                    },
+                    (None, None) => {
+                        let conf = if cluster.conf.skip_tls.unwrap_or(false) {
+                            ClusterConf::new_insecure(None, cluster.conf.server.clone())
+                        } else {
+                            ClusterConf::new(None, cluster.conf.server.clone())
+                        };
+                        cluster_map.insert(cluster.name.clone(), conf);
                     }
                 }
-                (None, Some(cert_data)) => match ::base64::decode(cert_data.as_str()) {
-                    Ok(mut cert) => {
-                        cert.retain(|&i| i != 0);
-                        let cert_pem = String::from_utf8(cert).map_err(|e| {
-                            KubeError::ConfigFileError(format!(
-                                "Invalid utf8 data in certificate: {}",
-                                e.description()
-                            ))
-                        })?;
-                        cluster_map.insert(
-                            cluster.name.clone(),
-                            ClusterConf::new(Some(cert_pem), cluster.conf.server),
-                        );
-                    }
-                    Err(e) => {
-                        println!(
-                            "Invalid certificate data, could not base64 decode: {}",
-                            e.description()
-                        );
-                    }
-                },
-                (None, None) => {
-                    let conf = if cluster.conf.skip_tls.unwrap_or(false) {
-                        ClusterConf::new_insecure(None, cluster.conf.server)
-                    } else {
-                        ClusterConf::new(None, cluster.conf.server)
-                    };
-                    cluster_map.insert(cluster.name.clone(), conf);
-                }
             }
         }
 
         // copy over contexts
         let mut context_map = HashMap::new();
-        for context in iconf.contexts.iter() {
-            context_map.insert(context.name.clone(), context.conf.clone());
+        for iconf in iconfs.iter() {
+            for context in iconf.contexts.iter() {
+                context_map.insert(context.name.clone(), context.conf.clone());
+            }
         }
 
         // copy over users
         let mut user_map = HashMap::new();
-        for user in iconf.users.into_iter() {
-            user_map.insert(user.name.clone(), user.conf.into());
+        for iconf in iconfs.iter() {
+            for user in iconf.users.iter() {
+                user_map.insert(user.name.clone(), user.conf.clone().into());
+            }
         }
 
+        let sources = env::join_paths(paths.into_iter())?.into_string()?;
```

Comment:
Can we handle the into_string here without a `?`?

This makes the compiler tell you implement `std::convert::From::from` from `OsString` for `KubeError`, but that's because `into_string` returns the `OsString` you called it with in the `Err` variant if it fails.  I don't really like having that conversion since it doesn't really semantically make sense to say I can go from an `OsString` to a `KubeError`.  This is also only used for display, and I feel like it's pretty unlikely you'll have non-utf8 characters in the path (I think other things will break in that case).  I think something like this should work:

```rust
        let sources = match env::join_paths(paths.into_iter())?.into_string() {
            Ok(srcs) => srcs,
            Err(_) => "[config paths contain non-utf8 characters, cannot be displayed]".to_string()
        };
```

Then we can get rid of all the `OsString` stuff in `error.rs`

Reply:
Absolutely!